source/agora/consensus/EnrollmentManager.d

```diff
@@ -797,6 +797,53 @@ public class EnrollmentManager
         return true;
     }
 
+    /***************************************************************************
+
+        Get the pre-image hash for this validator
+
+        Params:
+            enroll_key = UTXO enrollemnt key
+            height = the block height of pre-image to be returned
+
+        Returns:
+            The preimage hash if found otherwise Hash.init
+
+    ***************************************************************************/
+
+    public Hash getPreimage (in Height height) @safe nothrow
```

Comment:
I don't think you need this function having the parameter, `height` for block signing. Instead, you can use the following two functions. The nominator already has a `KeyPair` object so the second `getPreimage` can be used for that.
The name of `getValidatorPreImageAt` would be better because it already has `getValidatorPreimage` function.

Reply:
But if we have the cycle in memory then isn't it more efficient to fetch it from there rather than finding the utxo from our public key and then querying the database?

Reply:
Right, it is. But I think the three overridden functions should do a similar operation, which keeps other developers from confusing the code. If you prefer your way, I would recommend changing the second and third functions to `getValidatorPreimageAt`.

Reply:
Updated

Reply:
I would suggest renaming the function again. There are so many `getPreimage`-like functions. It's because the `EM` manages its own enrollment and the validator set. I think we need to make the names of the functions clearer. How about changing it to `getThisValidatorPreimage` or `getThisPreimage`?

Reply:
Good!